// ajax/lib/Ajax.js
var $ = require('jQuery');

const HTTP_POST 	= 'POST';
const HTTP_PUT 		= 'PUT';
const HTTP_GET 		= 'GET';
const HTTP_DELETE   = 'DELETE';
const CONTENT_TYPE  = 'application/json; charset=utf-8';

/* private */ function send(method, url, obj, success, failure, complete)
{
	$.ajax(
        {
            url 		: url,
            cache 		: false,
            type 		: method,
            contentType : CONTENT_TYPE,
            data		: JSON.stringify( obj )
        	/*
            beforeSend  : function (request)
            {
            	// Add custom headers here :
                request.setRequestHeader("myheader", "myvalue");
            },
            */
        }
    )
    .fail(
        function (xhr, textStatus, err) 
        {
        	if (failure)
        	{
        		failure(err, obj, xhr, textStatus);
        	}
    	}
	)
	.success(
        function (xhr, textStatus) 
        {
            if (success != null) {
                success(xhr);
            }
                
        }
	)
    .complete(
        function (xhr, textStatus) {
        	if (complete != null)
        		complete(xhr, textStatus);
        }
    );
}

/* class */ function Ajax()
{
	/* public async void */
	this.post = function(url, obj, success, failure, complete)
	{
		send(HTTP_POST, url, obj, success, failure, complete);
	};

	/* public async void */
	this.get = function(url, obj, success, failure, complete)
	{
		send(HTTP_GET, url, obj, success, failure, complete);
	};

	/* public async void */
	this.delete = function(url, obj, success, failure, complete)
	{
		send(HTTP_DELETE, url, obj, success, failure, complete);
	};

	/* public async void */
	this.put = function(url, obj, success, failure, complete)
	{
		send(HTTP_PUT, url, obj, success, failure, complete);
	};
}

module.exports = new Ajax();
