// observables/lib/observables.js
const EventEmitter = require('events').EventEmitter;

/* Inherits */ ObservableCollection.prototype = 
/* From */ 	   EventEmitter.prototype;
function ObservableCollection(of)
{
	/* Constructor */
	EventEmitter.call(this);

	var items = new Array();

	this.add = function(item)
	{
		if (!(item instanceof of))
			throw 'Invalid item added to the collection';

		this.emit('adding', item);
		items.push(item);
		this.emit('added', item);
	};

	this.remove = function(index)
	{
		var item = this.items[index];
		this.emit('removing', item);
		items.splice(index);
		this.emit('removed', item);
	};

	this.set = function(item, index)
	{
		if (!(item instanceof of))
			throw 'Invalid item set in the collection';

		this.emit('setting', item);
		items[index] = item;
		this.emit('set', item);
	};
    
    this.get = function(index)
    {
        return items[index];
    };
    
    this.forEach = function(callback)
    {
        return items.forEach(callback);
    };
    
    this.toArray = function()
    {
        return items;
    }
}

/* Inherits */ ObservableDictionary.prototype = 
/* From */ 	   EventEmitter.prototype;
function ObservableDictionary(of)
{
	/* Constructor */
	EventEmitter.call(this);

	this.add = function(key, item)
	{
		if (!(item instanceof of))
			throw 'Invalid item added to the collection';
        if (this[key])
            return false;
            
		this.emit('adding', item);
		this[key] = item;
		this.emit('added', item);
        
        return true;
	};

	this.clear = function()
	{
		for(key in this)
		{
			var obj = this[key];
			if (obj instanceof of)
				delete this[key];
		}
	};

	this.remove = function(key)
	{
		var item = this[key];
		if (item == null)
			return false;

		this.emit('removing', item);
		delete this[key];
		this.emit('removed', item);
		return true;
	};

	this.set = function(item, key)
	{
		if (!(item instanceof of))
			throw 'Invalid item set in the collection';

		this.emit('setting', item);
		this[key] = item;
		this.emit('set', item);
	};

	this.forEach = function(callback)
	{
		for(key in this)
		{
			var obj = this[key];
			if (obj instanceof of)
				callback(obj, key);
		}
	};
	
	this.toArray = function()
	{	
		var result = [];

		for(key in this)
		{
			var obj = this[key];
			if (obj instanceof of)
				result.push(obj);
		} 
			

		return result;
	}
}

module.exports = {
    Collection : ObservableCollection,
    Dictionary : ObservableDictionary
};
